File: src/struct_frame/boilerplate/rust/frame_base.rs
// Frame Base - Core utilities for struct-frame (Rust)
// Mirrors frame_base.hpp from C++ boilerplate

/// Checksum result - two bytes
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub struct FrameChecksum {
    pub byte1: u8,
    pub byte2: u8,
}

/// Calculate Fletcher-16 checksum over the given data slice.
///
/// # Arguments
/// * `data` - Buffer to checksum
/// * `magic1` - First magic number (mixed in after data)
/// * `magic2` - Second magic number (mixed in after magic1)
pub fn fletcher_checksum(data: &[u8], magic1: u8, magic2: u8) -> FrameChecksum {
    let mut b1: u8 = 0;
    let mut b2: u8 = 0;
    for &byte in data {
        b1 = b1.wrapping_add(byte);
        b2 = b2.wrapping_add(b1);
    }
    b1 = b1.wrapping_add(magic1);
    b2 = b2.wrapping_add(b1);
    b1 = b1.wrapping_add(magic2);
    b2 = b2.wrapping_add(b1);
    FrameChecksum { byte1: b1, byte2: b2 }
}

/// Extension-aware Fletcher-16 checksum.
/// Computes: Fletcher(base_data) -> mix magic1/magic2 -> Fletcher(ext_data).
/// When ext_data is empty the result is identical to fletcher_checksum.
///
/// # Arguments
/// * `base_data` - Non-extension portion of the payload
/// * `ext_data` - Extension portion of the payload (may be empty)
/// * `magic1` - First magic number (mixed in after base_data)
/// * `magic2` - Second magic number (mixed in after magic1)
pub fn fletcher_checksum_ext(base_data: &[u8], ext_data: &[u8], magic1: u8, magic2: u8) -> FrameChecksum {
    let mut b1: u8 = 0;
    let mut b2: u8 = 0;
    for &byte in base_data {
        b1 = b1.wrapping_add(byte);
        b2 = b2.wrapping_add(b1);
    }
    b1 = b1.wrapping_add(magic1);
    b2 = b2.wrapping_add(b1);
    b1 = b1.wrapping_add(magic2);
    b2 = b2.wrapping_add(b1);
    for &byte in ext_data {
        b1 = b1.wrapping_add(byte);
        b2 = b2.wrapping_add(b1);
    }
    FrameChecksum { byte1: b1, byte2: b2 }
}

/// Parser status indicating the reason a FrameMsgInfo is not valid.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub enum FrameMsgStatus {
    #[default]
    None = 0,
    WaitingForStart = 1,
    Collecting = 2,
    CrcFailure = 3,
    SyncRecovery = 4,
}

/// Result from frame parsing.
#[derive(Debug, Clone, Default)]
pub struct FrameMsgInfo {
    pub valid: bool,
    pub msg_id: u16,
    pub msg_len: usize,
    pub frame_size: usize,
    pub package_id: u8,
    pub sequence: u8,
    pub system_id: u8,
    pub component_id: u8,
    /// The raw message payload bytes (excluding framing overhead).
    /// Use `get_payload_span()` for a zero-copy slice or `extract_payload()` to move the data out.
    pub msg_data: Vec<u8>,
    /// Status indicating the reason this result is not valid (only meaningful when valid is false).
    pub status: FrameMsgStatus,
}

impl FrameMsgInfo {
    pub fn invalid() -> Self {
        Self::default()
    }

    /// Return a zero-copy slice of the payload bytes.
    #[inline]
    pub fn get_payload_span(&self) -> &[u8] {
        &self.msg_data
    }

    /// Move the payload bytes out, consuming this field's allocation.
    #[inline]
    pub fn extract_payload(self) -> Vec<u8> {
        self.msg_data
    }
}

/// Diagnostic counters for AccumulatingReader (stream mode).
///
/// All counters accumulate over the reader's lifetime.
/// Call `reset_diagnostics()` to clear them.
#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
pub struct ParserDiagnostics {
    /// Number of complete frames received with a bad CRC.
    /// Indicates noise or corruption on the line.
    pub cnt_crc_failures: u32,

    /// Number of times the parser discarded bytes and re-searched for a frame start.
    /// Indicates lost bytes or buffer overflows.
    pub cnt_sync_recoveries: u32,

    /// Total number of bytes discarded during sync recovery.
    /// Complements cnt_sync_recoveries with a byte-level view of data loss.
    pub cnt_failed_bytes: u32,

    /// Number of frames where the header length field does not match the expected
    /// message-struct size from the message-info callback.
    /// Vital for detecting mismatched definitions on profiles that carry an explicit length.
    pub cnt_len_errors: u32,

    /// Number of sequence-number gaps detected.
    /// Only incremented on profiles that carry a sequence field (e.g. ProfileNetwork).
    /// Indicates dropped packets.
    pub cnt_seq_gaps: u32,
}

/// Message metadata used for parsing (size + magic numbers)
#[derive(Debug, Clone, Copy)]
pub struct MessageInfo {
    pub size: usize,
    pub magic1: u8,
    pub magic2: u8,
    pub base_size: usize,  // Non-extension portion size (== size when no extensions)
    /// Minimum valid wire payload length (< size only for variable messages).
    pub min_size: usize,
}

impl MessageInfo {
    pub fn new(size: usize, magic1: u8, magic2: u8) -> Self {
        MessageInfo { size, magic1, magic2, base_size: size, min_size: size }
    }

    pub fn new_with_base_size(size: usize, magic1: u8, magic2: u8, base_size: usize) -> Self {
        MessageInfo { size, magic1, magic2, base_size, min_size: size }
    }

    pub fn new_variable(size: usize, magic1: u8, magic2: u8, base_size: usize, min_size: usize) -> Self {
        MessageInfo { size, magic1, magic2, base_size, min_size }
    }
}

/// Trait for messages that can be packed/unpacked to/from bytes
pub trait StructFrameMessage {
    const MSG_ID: u16;
    const MAX_SIZE: usize;
    const MAGIC1: u8;
    const MAGIC2: u8;
    /// Non-extension portion size. Equal to MAX_SIZE for messages with no extensions.
    const BASE_SIZE: usize = Self::MAX_SIZE;
    /// True if the message uses variable-length encoding for bounded fields.
    const IS_VARIABLE: bool;

    /// Serialize using variable-length encoding (actual data only for bounded fields).
    fn pack(&self, buf: &mut [u8]) -> usize;
    /// Serialize using fixed-size encoding (always MAX_SIZE bytes).
    fn pack_max_size(&self, buf: &mut [u8]) -> usize;
    /// Deserialize. For variable messages dispatches based on buf.len().
    fn unpack(buf: &[u8]) -> Option<Self>
    where
        Self: Sized;

    fn message_info() -> MessageInfo {
        MessageInfo::new(Self::MAX_SIZE, Self::MAGIC1, Self::MAGIC2)
    }
}

File: src/struct_frame/boilerplate/py/frame_base.py
# Frame Base - Core utilities for frame parsing (Python)
# Mirrors frame_base.hpp from C++ boilerplate

from typing import Union, List, Tuple, Optional
from dataclasses import dataclass
from enum import Enum


# =============================================================================
# Checksum
# =============================================================================

@dataclass
class FrameChecksum:
    """Checksum result - two bytes"""
    byte1: int = 0
    byte2: int = 0
    
    def __iter__(self):
        """Allow unpacking: b1, b2 = checksum"""
        return iter((self.byte1, self.byte2))
    
    def as_tuple(self) -> Tuple[int, int]:
        """Return as tuple"""
        return (self.byte1, self.byte2)


def fletcher_checksum(data: Union[bytes, List[int]], start: int = 0, end: Optional[int] = None, 
                      init1: int = 0, init2: int = 0) -> FrameChecksum:
    """
    Calculate Fletcher-16 checksum over the given data.
    
    Args:
        data: Buffer to checksum
        start: Start index (inclusive)
        end: End index (exclusive), defaults to len(data)
        init1: Magic number 1 (added at the end)
        init2: Magic number 2 (added at the end)
    
    Returns:
        FrameChecksum with byte1 and byte2
    """
    if end is None:
        end = len(data)
    
    byte1 = 0
    byte2 = 0
    for i in range(start, end):
        byte1 = (byte1 + data[i]) & 0xFF
        byte2 = (byte2 + byte1) & 0xFF
    
    # Add magic numbers at the end
    byte1 = (byte1 + init1) & 0xFF
    byte2 = (byte2 + byte1) & 0xFF
    byte1 = (byte1 + init2) & 0xFF
    byte2 = (byte2 + byte1) & 0xFF
    
    return FrameChecksum(byte1, byte2)


def fletcher_checksum_ext(data: Union[bytes, List[int]], start: int, base_end: int, end: int,
                          init1: int = 0, init2: int = 0) -> FrameChecksum:
    """
    Extension-aware Fletcher-16 checksum.

    Computes: Fletcher(data[start:base_end]) -> mix magic1/magic2 -> Fletcher(data[base_end:end]).
    When base_end == end the result is identical to fletcher_checksum.

    Args:
        data: Buffer to checksum
        start: Start index (inclusive)
        base_end: End of non-extension portion (exclusive); magic bytes are mixed here
        end: End index (exclusive)
        init1: Magic number 1
        init2: Magic number 2
    """
    byte1 = 0
    byte2 = 0
    for i in range(start, base_end):
        byte1 = (byte1 + data[i]) & 0xFF
        byte2 = (byte2 + byte1) & 0xFF
    byte1 = (byte1 + init1) & 0xFF
    byte2 = (byte2 + byte1) & 0xFF
    byte1 = (byte1 + init2) & 0xFF
    byte2 = (byte2 + byte1) & 0xFF
    for i in range(base_end, end):
        byte1 = (byte1 + data[i]) & 0xFF
        byte2 = (byte2 + byte1) & 0xFF
    return FrameChecksum(byte1, byte2)


# =============================================================================
# Parse Result
# =============================================================================

class FrameMsgStatus(Enum):
    """
    Reason code carried in FrameMsgInfo when valid is False.

    This lets callers distinguish between "parser is still accumulating bytes"
    (normal) and the various error conditions that cause the parser to discard
    data or fail validation.

    NONE              - Default / unset.  Used by one-shot buffer parsers when
                        no specific reason is available.
    WAITING_FOR_START - Parser is idle and searching for a start byte.  No
                        frame has been started yet (or the previous frame was
                        completed / reset).
    COLLECTING        - A frame is in progress; the parser is still accumulating
                        header or payload bytes.  More data is needed.
    CRC_FAILURE       - A complete frame was received but its CRC did not match.
                        Indicates noise or data corruption on the link.
    SYNC_RECOVERY     - The parser discarded one or more bytes to re-find a
                        valid frame start.  Triggered by bad bytes, unknown
                        message IDs, or internal buffer overflows.
    """
    NONE = 0
    WAITING_FOR_START = 1
    COLLECTING = 2
    CRC_FAILURE = 3
    SYNC_RECOVERY = 4


@dataclass
class FrameMsgInfo:
    """
    Result from frame parsing.
    
    Mirrors C++ FrameMsgInfo structure for compatibility.
    """
    valid: bool = False
    msg_id: int = 0           # Message ID (16-bit for extended profiles)
    msg_len: int = 0          # Payload length (message data only)
    frame_size: int = 0       # Total frame size (header + payload + footer)
    msg_data: bytes = b''     # Pointer to message data
    
    # Optional extended fields (for profiles that support them)
    package_id: int = 0
    sequence: int = 0
    system_id: int = 0
    component_id: int = 0

    # Parser state / reason code (meaningful when valid is False)
    status: FrameMsgStatus = FrameMsgStatus.NONE
    diagnostics: Optional['ParserDiagnostics'] = None
    
    def __bool__(self) -> bool:
        """Allow use in boolean context: while (result := reader.next()): ..."""
        return self.valid


# =============================================================================
# Parser Diagnostics
# =============================================================================

@dataclass
class ParserDiagnostics:
    """
    Error counters for parser diagnostics.

    Tracks error events detected by the AccumulatingReader in stream mode.
    These counters accumulate over the reader's lifetime and can be reset
    with reset_diagnostics().

    Attributes:
        cnt_crc_failures:   Complete frames where CRC validation failed.
                            Indicates noise or corruption on the line.
        cnt_sync_recoveries: Times the parser had to discard bytes and
                            re-search for a frame start.  Indicates lost
                            bytes or buffer overflows.
        cnt_failed_bytes:   Total bytes discarded when failures forced the
                    parser to restart searching for frame start.
        cnt_len_errors:     Frames where the explicit length in the header
                            does not match the expected message-struct size
                            returned by get_message_info().  Vital for
                            detecting sender/receiver definition mismatches.
        cnt_seq_gaps:       Sequence-number gaps in received messages.
                            Only incremented on profiles that carry a
                            sequence field (e.g. ProfileNetwork).
                            Indicates dropped packets.
    """
    cnt_crc_failures: int = 0
    cnt_sync_recoveries: int = 0
    cnt_failed_bytes: int = 0
    cnt_len_errors: int = 0
    cnt_seq_gaps: int = 0


# =============================================================================
# Parser State (for streaming parsers)
# =============================================================================

class ParserState(Enum):
    """Parser state machine states"""
    LOOKING_FOR_START = 0
    GOT_BASIC_START = 1     # Got 0x90, waiting for payload type byte
    GOT_UBX_SYNC1 = 2       # Got 0xB5, waiting for 0x62
    PARSING_HEADER = 3      # Parsing payload header fields
    PARSING_PAYLOAD = 4     # Parsing message data
    PARSING_FOOTER = 5      # Parsing CRC
